aws_advanced_python_wrapper/limitless_connection_plugin.py

```diff
@@ -0,0 +1,543 @@
+                logger.error("LimitlessRouterServiceImpl.IncorrectConfiguration")
+                raise e
+            except AwsWrapperError:
+                continue
+
+            try:
+                context.set_connection(self._plugin_service.connect(selected_host_info, context.get_props()))
+                if context.get_connection() is not None:
+                    return
+
+            except Exception as e:
+                if self._is_login_exception(e):
+                    return
+                selected_host_info.set_availability(HostAvailability.UNAVAILABLE)
+                logger.debug("LimitlessRouterServiceImpl.FailedToConnectToHost", selected_host_info.host)
+
+        raise AwsWrapperError(Messages.get("LimitlessRouterService.MaxRetriesExceeded"))
+
+    def _synchronously_get_limitless_routers_with_retry(self, context: LimitlessConnectionContext) -> None:
+        logger.debug("LimitlessRouterServiceImpl.SynchronouslyGetLimitlessRouters")
+        retry_count = -1
+        max_retries = WrapperProperties.MAX_RETRIES.get_int(context.get_props())
+        retry_interval_ms = WrapperProperties.GET_ROUTER_RETRY_INTERVAL_MILLIS.get_float(context.get_props())
+        first_iteration = True
+        while first_iteration or retry_count < max_retries:
+            # Emulate do while loop
+            first_iteration = False
+            try:
+                self._synchronously_get_limitless_routers(context)
+                if context.get_limitless_routers() is not None or len(context.get_limitless_routers()) > 0:
+                    return
+
+                time.sleep(retry_interval_ms)
+            except InterruptedError as e:
+                raise AwsWrapperError(Messages.get("LimitlessRouterService.InterruptedSynchronousGetRouter"), e)
+            except Exception as e:
+                if self._is_login_exception(e):
+                    raise e
+            finally:
+                retry_count += 1
+
+        raise AwsWrapperError(Messages.get("LimitlessRouterService.NoRoutersAvailable"))
+
+    def _synchronously_get_limitless_routers(self, context: LimitlessConnectionContext) -> None:
+        cache_expiration_nano: int = WrapperProperties.MONITOR_DISPOSAL_TIME_MS.get_int(context.get_props()) * 1_000_000
+
+        lock = self._force_get_limitless_routers_lock_map.compute_if_absent(
+            self._plugin_service.host_list_provider.get_cluster_id(),
+            lambda _: RLock()
+        )
+        if lock is None:
+            raise AwsWrapperError(Messages.get("LimitlessRouterService.LockFailedToAcquire"))
+
+        lock.acquire()
+        try:
+            limitless_routers = self._limitless_router_cache.get(
+                self._plugin_service.host_list_provider.get_cluster_id())
+            if limitless_routers is not None and len(limitless_routers) != 0:
+                context.set_limitless_routers(limitless_routers)
+                return
+
+            if context.get_connection() is None:
```

Comment:
In JDBC we also check if the connection is closed, I'm not sure if this is necessary or not though